### ISSUE TITLE
sparse_lane_no_dense_assignment is RED on main: cofit_arrow.rs constructs dense SaeAssignment twice in production code (#985/E1)

### ISSUE THREAD

**Issue by SauersML** [SauersML/gam#2693]
`gam-sae::sparse_lane_no_dense_assignment::sparse_lane_constructs_no_dense_assignment` fails on
`origin/main`, in 23 ms, with:

```
the production sparse-code lane must construct ZERO dense SaeAssignments — that dense N×K routing
state is the certification / debug-and-research lane only (#985 / E1);
found 1 offender(s): ["cofit_arrow.rs: production code contains `SaeAssignment::`"]
```

The guard scans `crates/gam-sae/src/sparse_dict/*.rs` with `cfg(test)` regions stripped, for
`["SaeAssignment {", "SaeAssignment::"]`. The offending sites are real production code:

- `crates/gam-sae/src/sparse_dict/cofit_arrow.rs:212` — `SaeAssignment::from_blocks_with_mode(...)`
  inside `build_linear_cofit_term` (`:138`), reached from `pub fn cofit_linear_via_arrow` (`:109`);
- `:503` — `SaeAssignment::from_blocks_with_mode_and_manifolds(...)` inside
  `build_composed_cofit_term` (`:400`), reached from `pub fn cofit_composed_via_arrow` (`:356`).

Both build a **dense `(n, g)` logits array** (`Array2::<f64>::from_elem((n, g), OFF)`) and hand it
to `AssignmentMode::top_k_support`, which is exactly the dense `N×K` routing state #985/E1 says
this lane must not construct.

So this is a **genuine architectural violation on main**, not a stale guard. Two dispositions are
possible and I do not think it should be decided by whoever is next fixing an unrelated test:

1. **`cofit_arrow` is production** → the dense assignment must go; the routing is already known
   (`blocks`), so a sparse construction should be reachable without materializing `(n, g)`.
2. **`cofit_arrow` is the certification / debug-and-research lane** → then it is in the wrong
   directory, and the fix is to move it out of the scanned set rather than to widen the guard's
   needles. Widening the guard to permit the thing it guards would make it inert.

## A survey that is NOT yet evidence

`cofit_linear_via_arrow`, `cofit_composed_via_arrow` and `ArrowCofitConfig` are re-exported from
`sparse_dict/mod.rs:73` and have **zero callers anywhere in the repository** — nothing in
`crates/`, no example, no bench. That is *suggestive* of disposition 2, and it is deliberately not
being acted on here.

A caller `grep` is not a liveness proof in this repo: it misses construction sites, `#[path]`-
included modules, and cfg-gated reachability. A previous six-option "dead code" survey had five of
its six spot-checks refuted, and separately today one helper with "zero callers" turned out to have
four while another with a "single production caller" turned out to have none. **Anyone deleting
this must count constructor functions, brace-match the `cfg(test)` regions, check `#[path]`
inclusion, and verify against `git show origin/main:<file>` rather than the working tree** — the
worktree is shared and dirty. Then delete with that evidence in the commit message.

(#2660, the composed arrow-Schur co-fit idempotence certificates, is now closed — so the
certificates that used to justify this module's existence may no longer need it. That makes
disposition 2 more likely, and makes the constructor survey worth doing properly rather than
worth skipping.)

Filed rather than fixed: found while triaging an unrelated `gam-sae` test block, and it is a
design decision about a module I do not own. Not urgent — the guard is the only thing failing, and
it is failing correctly.

**Comment by SauersML** [comment]
## Re-verified RED, then fixed. Landed `55c60632051befe347d1d221f08d55a9d2754aad`.

### 1. Re-verification (this was not stale)

Base pinned once and reused in every arm: **`517b6303fdcc040dc7fba16ac8984ca97d4bce47`** (fresh
`origin/main`, ~44 commits after the filing). MSI job `14618138`, `cargo test -p gam-sae --test
sparse_lane_no_dense_assignment`:

```
test result: FAILED. 2 passed; 1 failed; 0 ignored; 0 measured; 0 filtered out; finished in 0.12s
test sparse_lane_constructs_no_dense_assignment ... FAILED
  ... found 1 offender(s): ["cofit_arrow.rs: production code contains `SaeAssignment::`"]
```

Still red, and correctly so.

### 2. Dividing out the number the message already carries

`found 1 offender(s)` is a **needle count, not a site count** — the loop pushes one string per
`(file, needle)` pair. Decomposed at the pinned base:

| needle | production region (L1..670) | `#[cfg(test)]` region (L671..1229) |
|---|---|---|
| `` `SaeAssignment::` `` | **2** (`:212` `from_blocks_with_mode`, `:503` `from_blocks_with_mode_and_manifolds`) | 0 |
| `` `SaeAssignment {` `` | 0 | 0 |

So: 1 offender string = 1 file x 1 needle = **2 real production construction sites**. The `1` was
never a count of the defect. (`SaeAssignmentState` on L20 contains the substring `SaeAssignment` but
matches neither needle, so it is not a false positive either.)

### 3. Which disposition — decided by measurement, not preference

**Disposition 1 is structurally impossible at this base.** The dense engine's *only* constructor is

```rust
// crates/gam-sae/src/manifold/construction.rs:780
pub fn new(atoms: Vec<SaeManifoldAtom>, assignment: SaeAssignment) -> Result<Self, String>
```

and `SaeAssignment` **stores** `logits: Array2<f64>` (`assignment.rs`) — the dense `N x K` state is a
field, not a transient buffer. Entering the dense engine therefore *forces* that state into
existence; there is no sparse in-core entry (that is the Stage 2 `SaeAssignmentState::from_topk_support`
seam, still specced to the engine lane on #2023 — `cofit_arrow.rs`'s own header says exactly this,
and calls its own scope "**Stage 1 ... DENSE assignment, moderate K**"). You cannot delete the dense
assignment from `cofit_arrow` while it calls `SaeManifoldTerm`; the routing being "already known"
does not help, because the engine's front door is typed on the dense object.

**So: disposition 2.** The guard's own doc names the legitimate home — *"that path lives in the
manifold engine, not here"* — and `sparse_dict/mod.rs` defines the lane as "**no dense `N x K` object
anywhere**". `cofit_arrow` is a bridge *from* a block routing *into* the dense certification engine.
It was on the wrong side of the boundary.

### 4. The liveness survey you asked for, done to your spec

Redone at the pinned base against `git cat-file -p BASE:<path>`, never the (dirty, shared) worktree.
`cofit_linear_via_arrow`, `cofit_composed_via_arrow`, `ArrowCofitConfig`, `ArrowCofitReport` and
`cofit_arrow` have **zero** references anywhere in the repo outside the file itself and the single
`mod cofit_arrow;` + `pub use cofit_arrow::{...}` pair in `sparse_dict/mod.rs`: no `#[path]`
inclusion, no second `mod` declaration, no constructor of either struct, no bench, no example, no
`gam-pyffi`. That survey is why the **move is safe and a deletion was not chosen**: the five
`#2023`/`#2397`/`#2660` certificates inside the file are the only consumers, and they are worth
keeping. I did not delete anything.

### 5. What landed

`crates/gam-sae/src/sparse_dict/cofit_arrow.rs` -> `crates/gam-sae/src/manifold/cofit_arrow.rs`,
verbatim apart from: two `use super::{block_chart, coordinate}` imports repointed at
`crate::sparse_dict`'s **public** surface, three doc links rewritten, and a header paragraph
recording the boundary argument above. `mod`/`pub use` moved from `sparse_dict/mod.rs` to
`manifold/mod.rs`.

**No bound was relaxed and no needle was widened.** The guard was *strengthened*, in the tightening
direction only:

* it now **recurses** `sparse_dict/**` — a no-op today (no subdirs), which is precisely why it was
  cheap to close now, and it forecloses "move it into a subdirectory" as a future escape;
* it adds **`SaeManifoldTerm`** as a third needle, gating the **mechanism** rather than the symptom:
  naming the dense engine aggregate in the sparse lane forces a dense assignment via its only
  constructor. `crate::manifold::realised_rank_charge_dof` (`block.rs:1963`, `block_stream.rs:946`)
  is deliberately **not** a needle — a pure helper that builds no routing state, and it stays legal;
* the assertion now prints the **scanned file count**, so a vacuous scan is no longer
  byte-identical to a clean one.

### 6. Measurements (MSI, base `517b6303f`, one clone, jobs `14618138` / `14618227`)

| arm | result |
|---|---|
| re-verify: shipped guard, unpatched | `test result: FAILED. 2 passed; 1 failed` — 1 offender |
| **ARM A (negative control)**: strengthened guard, **unpatched** | `test result: FAILED. 2 passed; 1 failed` — `scanned 18 file(s), found 2 offender(s)`: `SaeAssignment::` **and** `SaeManifoldTerm`. Both new needles proven able to fire. |
| **ARM B**: strengthened guard + the move | `test result: ok. 3 passed; 0 failed` |
| certificates at the new path | `test result: ok. 5 passed; 0 failed` (all five `#2023`/`#2397` tests) |
| `cargo check --workspace --exclude gam-pyffi --all-targets --locked` | rc=0 |
| `SaeAssignment`/`SaeManifoldTerm` left anywhere under `sparse_dict/` | **0** |

`scripts/ban_scanner.sh` positive control (it prints only violations, so a green run is
byte-identical to a non-run): a planted `#[allow(dead_code)]` in `manifold/cofit_arrow.rs` gives
exit **1** naming `crates/gam-sae/src/manifold/cofit_arrow.rs:1245`; removing it gives exit **0**
with **0** lines of output.

Landed by plumbing as `55c60632051befe347d1d221f08d55a9d2754aad` on parent `55c56d6f4`, gated on
`git diff-tree BASE PARENT` being empty for all five touched paths, and `git diff-tree PARENT COMMIT`
showing exactly `A manifold/cofit_arrow.rs`, `D sparse_dict/cofit_arrow.rs`, `M` on the two `mod.rs`
and the guard — no foreign hunks. Each landed blob re-read back from the commit and confirmed
byte-identical to what was tested.

Closing.
